app/ide-desktop/lib/dashboard/src/layouts/AssetVersions/AssetVersions.tsx

```diff
@@ -23,56 +39,108 @@ import type AssetTreeNode from '#/utilities/AssetTreeNode'
+          key: uniqueString.uniqueString(),
+          lastModified: dateTime.toRfc3339(new Date()),
+          versionId: variables.placeholderId,
+        },
+        ...oldVersions,
+      ])
+    },
+    onSuccess: async () => {
+      // `backend.restoreProject` does not return the ID of the new version, so a full refetch is
+      // necessary.
+      await versionsQuery.refetch()
+    },
+    onError: (error: unknown) => {
+      toastAndLog('restoreProjectError', error, item.item.title)
+    },
+    onSettled: (_data, _error, variables) => {
+      setPlaceholderVersions(oldVersions =>
+        oldVersions.filter(version => version.versionId !== variables.placeholderId)
+      )
+    },
+  })
 
   return (
     <div className="pointer-events-auto flex flex-1 shrink-0 flex-col items-center overflow-y-auto overflow-x-hidden">
-      {(() => {
-        if (!isCloud) {
-          return <div>{getText('localAssetsDoNotHaveVersions')}</div>
-        } else if (isPending) {
-          return <Spinner size={32} state={spinnerModule.SpinnerState.loadingMedium} />
-        } else if (status === 'error') {
-          return <div>{getText('listVersionsError')}</div>
-        } else if (versions.length === 0) {
-          return <div>{getText('noVersionsFound')}</div>
-        } else if (!latestVersion) {
-          return <div>{getText('fetchLatestVersionError')}</div>
-        } else {
-          return versions.map((version, i) => (
+      {!isCloud ? (
+        <div>{getText('localAssetsDoNotHaveVersions')}</div>
+      ) : versionsQuery.isPending ? (
+        <Spinner size={32} state={spinnerModule.SpinnerState.loadingMedium} />
+      ) : versionsQuery.isError ? (
+        <div>{getText('listVersionsError')}</div>
+      ) : versionsQuery.data.length === 0 ? (
+        <div>{getText('noVersionsFound')}</div>
+      ) : latestVersion == null ? (
+        <div>{getText('fetchLatestVersionError')}</div>
+      ) : (
+        [
+          ...placeholderVersions.map((version, i) => (
+            <AssetVersion
+              key={version.versionId}
+              placeholder
+              number={versionsQuery.data.length + placeholderVersions.length - i}
+              version={version}
+              item={item}
+              backend={backend}
+              latestVersion={latestVersion}
+              dispatchAssetListEvent={dispatchAssetListEvent}
+              doRestore={() => {}}
+            />
+          )),
+          ...versionsQuery.data.map((version, i) => (
             <AssetVersion
               key={version.versionId}
-              number={versions.length - i}
+              number={versionsQuery.data.length - i}
               version={version}
-              item={item.item}
+              item={item}
               backend={backend}
               latestVersion={latestVersion}
+              dispatchAssetListEvent={dispatchAssetListEvent}
+              doRestore={() => {
+                restoreMutation.mutate({
```

Comment:
can we use `mutateAsync` here? If we pass a function that returns a promise to a button, button will automatically set loading state and remove it once promise resolves